### DISCUSSION TITLE
Export nodal data to VTK format

### DISCUSSION THREAD

**Discussion by CsatiZoltan** [neperfepx/neper#253] [Q&A] [answered]
Hi,

I tried to plot the displacement field in Neper. What I realized was that using the `-datanodecol` option only, without providing element visualization using for example the `-dataelt3dcol` option, the PNG output defaults to the colour of the element sets. It is possible to plot the displacement field by interpolating the nodal values on the elements, so
```
-dataelt3dcol from_nodes
-datanodecol disp1
```
works fine for the PNG output. On the other hand, no (nodal or interpolated element) data is written to the VTK file, when I specify the image format to be `vtk`.

How can I export the displacement field into a VTK file? The same observations apply to other nodal data as well, such as the coordinates.

Thank you,
Zoltan

**EDIT**: If nodal data could be output into VTK format, that would allow one to export the coordinates as well, hence supporting the visualization of the deformed configuration.

**Comment by rquey** [comment, marked as answer]
> Hi,
> 
> I tried to plot the displacement field in Neper. What I realized was that using the `-datanodecol` option only, without providing element visualization using for example the `-dataelt3dcol` option, the PNG output defaults to the colour of the element sets.

Yes, because `-datanodecol` defines the colors of the nodes, not the colors of the elements. You could see the (colored) nodes using `-shownode 1` (and probably `-datanoderad` to increase the node radii).

> It is possible to plot the displacement field by interpolating the nodal values on the elements, so
> 
> ```
> -dataelt3dcol from_nodes
> -datanodecol disp1
> ```

Yes.

> works fine for the PNG output. On the other hand, no (nodal or interpolated element) data is written to the VTK file, when I specify the image format to be `vtk`.

Indeed, no "color data" are written to the VTK file at the moment.

> How can I export the displacement field into a VTK file? The same observations apply to other nodal data as well, such as the coordinates.
> 
> Thank you,
> Zoltan
> 
> **EDIT**: If nodal data could be output into VTK format, that would allow one to export the coordinates as well, hence supporting the visualization of the deformed configuration.

You can use `-datanodecoo coo`, which plots the deformed mesh in PNG and writes a `VECTORS coo` data field in the VTK file.

**Reply by CsatiZoltan** [reply]
> You can use `-datanodecoo coo`, which plots the deformed mesh in PNG and writes a `VECTORS coo` data field in the VTK file.

Yes, I already used that option. The deformed mesh is shown in the PNG output, but not in the VTK file. Perhaps because the magnification factor is not taken into account for the VTK output?

**Reply by rquey** [reply]
I don't use the VTK output, but I think that the coo data are not used by default, to deform the mesh - you may have to "load" them. And you are right the magnification factor is not taken into account in the VTK output.

**Reply by CsatiZoltan** [reply]
You were right, I had to first [select the vtk file](https://discourse.paraview.org/t/ive-just-opened-my-dataset-i-cant-add-filters-and-everything-is-greyed-out/4511) to enable the filters (otherwise all of them are disabled!), and then apply `coo` for the [Warp By Vector](https://docs.paraview.org/en/latest/UsersGuide/filteringData.html#warp-by-vector) filter.

While exporting custom nodal data (such as the `disp` output) to VTK is not yet supported by Neper, your method to plot the deformed configuration works and that was my original intent. I cannot mark your answer as "Accepted", so please do that yourself. Thank you for the solution.

**Reply by CsatiZoltan** [reply]
@rquey Although I still think it would be useful to add the feature of plotting nodal data (what you called "color data"). I am not pushing it from my side, it is just handy if one could superimpose the original and deformed meshes in Paraview. That would be more elegant than exporting `coo` from step $n$ and from step $0$ into two .vtk files, and then subtract them using filters.

**Reply by rquey** [reply]
So, in this case, you would want the `disp` data in the vtk file, is that it?

**Reply by CsatiZoltan** [reply]
@rquey Yes.

**Reply by rquey** [reply]
@CsatiZoltan The `-datanodecol` data are now written to the vtk file (version `4.2.1-6`). They can be of any type (not only scalars), so this will work with `disp`. If not scalars, you have to disable the png output (`-imageformat vtk`).

**Reply by CsatiZoltan** [reply]
Works perfectly! Thank you.
